src/cluster_slot_stats.c

```diff
@@ -121,12 +125,60 @@ static void addReplyOrderBy(client *c, int order_by, long limit, int desc) {
     addReplySortedSlotStats(c, slot_stats, limit);
 }
 
+static int canAddNetworkBytes(client *c) {
+    /* First, cluster mode must be enabled.
+     * Second, command should target a specific slot.
+     * Third, blocked client is not aggregated, to avoid duplicate aggregation upon unblocking. */
+    return server.cluster_enabled && server.cluster_slot_stats_enabled && c->slot != -1 && !(c->flag.blocked);
+}
+
+/* Resets applicable slot statistics. */
+void clusterSlotStatReset(int slot) {
+    /* key-count is exempt, as it is queried separately through countKeysInSlot(). */
+    server.cluster->slot_stats[slot].network_bytes_in = 0;
+}
+
+void clusterSlotStatResetAll(void) {
+    if (server.cluster == NULL) return;
+
+    memset(server.cluster->slot_stats, 0, sizeof(server.cluster->slot_stats));
+}
+
+/* Adds network ingress bytes of the current command in execution,
+ * calculated earlier within networking.c layer.
+ *
+ * Note: Below function should only be called once c->slot is parsed.
+ * Otherwise, the aggregation will be skipped due to canAddNetworkBytes() check failure.
+ * */
+void clusterSlotStatsAddNetworkBytesIn(client *c) {
+    if (!canAddNetworkBytes(c)) return;
+
+    if (c->cmd->proc == execCommand) {
+        /* Accumulate its corresponding MULTI RESP; *1\r\n$5\r\nmulti\r\n */
+        c->net_input_bytes_curr_cmd += 15;
+    }
+
+    server.cluster->slot_stats[c->slot].network_bytes_in += c->net_input_bytes_curr_cmd;
+}
+
+/* Adds network ingress bytes from sharded pubsub subscription.
+ * Since sharded pubsub targets a specific slot, we are able to aggregate its ingress bytes under per-slot context. */
+void clusterSlotStatsAddNetworkBytesInForShardedPubSub(robj *channel, robj *message) {
+    int slot = keyHashSlot(channel->ptr, sdslen(channel->ptr));
```

Comment:
It's not, because this is on the input side. We are computing the hash twice though, once inside the sharded pubsub function processing and once here. We should be able to merge the two invocations.

Reply:
1 seems OK to me.